cmd/thermal-recorder/frameloop.go

```diff
@@ -56,16 +61,24 @@ func (fl *FrameLoop) Current() *lepton3.Frame {
 	return fl.frames[fl.currentIndex]
 }
 
-// Previous returns the previous frame.
-// Note: data on all returned frame objects will eventually get overwritten
+// Previous returns a copy of the previous frame.
 func (fl *FrameLoop) Previous() *lepton3.Frame {
+	fl.mu.Lock()
+	defer fl.mu.Unlock()
+	if fl == nil {
```

Comment:
Why would fl be nil? This can only happen if you do something like this:

```go
var fl *FrameLoop
fl.Previous()
```

This is pretty unlikely.

Reply:
If you mean to return nil when there's no previous frame yet you have to check `fl.currentIndex` and `fl.bufferFull`.

Something like: `if fl.bufferFull || fl.currentIndex != 0 {`

Reply:
If the TakeSnapshot D-Bus call is made between thermal-recorder starting and it connecting to the camera the first time fl can be nil (this can take a while some times). But because of this the mu.Lock should be moved to after the null check.